osd: 'osd tree in|out|up|down' to filter tree results

liewegas opened the pull request:
This lets you easily see down OSDs and where they are with
'ceph osd tree down'.

<pre>
gnit:build (wip-osd-tree) 11:37 AM $ bin/ceph osd tree
ID WEIGHT  TYPE NAME     UP/DOWN REWEIGHT PRIMARY-AFFINITY 
-1 5.00000 root default                                    
-2 3.00000     host gnit                                   
 0 1.00000         osd.0      up  1.00000          1.00000 
 1 1.00000         osd.1      up  1.00000          1.00000 
 2 1.00000         osd.2      up        0          1.00000 
-3 2.00000     host foo                                    
 3 1.00000         osd.3    down        0          1.00000 
 4 1.00000         osd.4      up  1.00000          1.00000 
gnit:build (wip-osd-tree) 11:37 AM $ bin/ceph osd tree up
ID WEIGHT  TYPE NAME     UP/DOWN REWEIGHT PRIMARY-AFFINITY 
-1 5.00000 root default                                    
-2 3.00000     host gnit                                   
 0 1.00000         osd.0      up  1.00000          1.00000 
 1 1.00000         osd.1      up  1.00000          1.00000 
 2 1.00000         osd.2      up        0          1.00000 
-3 2.00000     host foo                                    
 4 1.00000         osd.4      up  1.00000          1.00000 
gnit:build (wip-osd-tree) 11:37 AM $ bin/ceph osd tree down
ID WEIGHT  TYPE NAME     UP/DOWN REWEIGHT PRIMARY-AFFINITY 
-1 5.00000 root default                                    
-3 2.00000     host foo                                    
 3 1.00000         osd.3    down        0          1.00000 
gnit:build (wip-osd-tree) 11:37 AM $ bin/ceph osd tree in
ID WEIGHT  TYPE NAME     UP/DOWN REWEIGHT PRIMARY-AFFINITY 
-1 5.00000 root default                                    
-2 3.00000     host gnit                                   
 0 1.00000         osd.0      up  1.00000          1.00000 
 1 1.00000         osd.1      up  1.00000          1.00000 
-3 2.00000     host foo                                    
 4 1.00000         osd.4      up  1.00000          1.00000 
gnit:build (wip-osd-tree) 11:37 AM $ bin/ceph osd tree out
ID WEIGHT  TYPE NAME     UP/DOWN REWEIGHT PRIMARY-AFFINITY 
-1 5.00000 root default                                    
-2 3.00000     host gnit                                   
 2 1.00000         osd.2      up        0          1.00000 
-3 2.00000     host foo                                    
 3 1.00000         osd.3    down        0          1.00000 
gnit:build (wip-osd-tree) 11:37 AM $ bin/ceph osd tree out down
ID WEIGHT  TYPE NAME     UP/DOWN REWEIGHT PRIMARY-AFFINITY 
-1 5.00000 root default                                    
-3 2.00000     host foo                                    
 3 1.00000         osd.3    down        0          1.00000 
gnit:build (wip-osd-tree) 11:37 AM $ bin/ceph osd tree out up
ID WEIGHT  TYPE NAME     UP/DOWN REWEIGHT PRIMARY-AFFINITY 
-1 5.00000 root default                                    
-2 3.00000     host gnit                                   
 2 1.00000         osd.2      up        0          1.00000 
gnit:build (wip-osd-tree) 11:37 AM $ bin/ceph osd tree in out
Error EINVAL: cannot specify both up and down or both in and out
gnit:build (wip-osd-tree) 11:37 AM $ bin/ceph osd tree up down
Error EINVAL: cannot specify both up and down or both in and out
gnit:build (wip-osd-tree) 11:37 AM $ bin/ceph osd tree foo
foo not valid:  foo not in up|down|in|out
Invalid command:  unused arguments: [u'foo']
osd tree {<int[0-]>} {up|down|in|out [up|down|in|out...]} :  print OSD tree
Error EINVAL: invalid command
gnit:build (wip-osd-tree) 11:38 AM $ bin/ceph osd tree foo up
foo not valid:  foo not in up|down|in|out
Invalid command:  unused arguments: [u'foo', u'up']
osd tree {<int[0-]>} {up|down|in|out [up|down|in|out...]} :  print OSD tree
Error EINVAL: invalid command
</pre>

branch-predictor commented on src/osd/OSDMap.cc:
```diff
@@ -2849,8 +2849,34 @@ void OSDMap::print(ostream& out) const
+  bool down = false;
+  bool up = false;
+  bool in = false;
+  bool out = false;
```
save 3 bytes by putting these bools in a bitfields.

liewegas replied:
this was less code that defining bitfields; not worth optimizing memory usage for a one-off thing like this.

branch-predictor commented on src/osd/OSDMap.cc:
```diff
@@ -2849,8 +2849,34 @@ void OSDMap::print(ostream& out) const
+    : Parent(crush), osdmap(osdmap_) {
+    if (state.find("up") != std::string::npos)
+      up = true;
+    else if (state.find("down") != std::string::npos)
```
It would be even more useful if it would be possible to combine flags (like "in+down"). Right now it accepts only one flag.

liewegas replied:
you actually can do that.. that's why it's find() and not operator==.  'ceph osd tree out,down' or 'out+down' or even 'downandout' works (but sadly not 'downandoutinbeverlyhills' as 'in' is in there too)

liewegas replied:
oh nevermind, the MOnCommands doesn't allow that.  i'll fix that up.

branch-predictor commented on src/osd/OSDMap.cc:
```diff
@@ -2906,16 +2933,41 @@ class OSDTreeFormattingDumper : public CrushTreeDumper::FormattingDumper {
+  OSDTreeFormattingDumper(const CrushWrapper *crush, const OSDMap *osdmap_,
+			  const string& state)
+    : Parent(crush), osdmap(osdmap_) {
+    if (state.find("up") != std::string::npos)
```
It's the same code as above, how about moving it to single, separate func that returns int containing flags, then passing that around?

branch-predictor commented on src/tools/osdmaptool.cc:
```diff
@@ -619,12 +619,12 @@ int main(int argc, const char **argv)
     if (tree_formatter) {
       tree_formatter->open_object_section("tree");
-      osdmap.print_tree(tree_formatter.get(), NULL);
+      osdmap.print_tree(tree_formatter.get(), NULL, string());
```
That string construction wouldn't be necessary if it would accept int.

liewegas commented:
ok, fixed everything and added tests, except for the common parent for the dumper class.  That's not possible without some pretty multi-inheritance because the (immediate) parent classes are different and it would probably end up being more code than the two trivialish overrides.

liewegas commented:
http://pulpito.ceph.com/sage-2017-05-29_22:34:44-rados-wip-sage-testing---basic-smithi/

branch-predictor commented:
LGTM

```
Reviewed-By: Piotr Dałek <piotr.dalek@corp.ovh.com>
```

tchaikov commented on src/osd/OSDMap.cc:
```diff
@@ -2900,22 +2917,38 @@ class OSDTreePlainDumper : public CrushTreeDumper::Dumper<TextTable> {
 
 private:
   const OSDMap *osdmap;
+  unsigned filter;
```
could make this a `const`, as we don't change it after initializing it.

tchaikov commented on src/crush/CrushTreeDumper.h:
```diff
@@ -76,8 +76,31 @@ namespace CrushTreeDumper {
       clear();
     }
 
+    virtual bool should_dump_leaf(int i) {
```
could be a `const`.

tchaikov commented on src/crush/CrushTreeDumper.h:
```diff
@@ -76,8 +76,31 @@ namespace CrushTreeDumper {
+    virtual bool should_dump_leaf(int i) {
+      return true;
+    }
+    virtual bool should_dump_empty_bucket() {
```
ditto.
